stan/math/rev/mat/fun/cholesky_decompose.hpp

```diff
@@ -223,6 +258,135 @@ class cholesky_scalar : public vari {
+            A.rows() * (A.rows() + 1) / 2)),
+        variRefL_(ChainableStack::instance().memalloc_.alloc_array<vari*>(
+            A.rows() * (A.rows() + 1) / 2)) {
+    size_t pos = 0;
+    for (size_type j = 0; j < M_; ++j) {
+      for (size_type i = j; i < M_; ++i) {
+        variRefA_[pos] = A.coeffRef(i, j).vi_;
+        variRefL_[pos] = new vari(L_A.coeffRef(i, j), false);
+        ++pos;
+      }
+    }
+  }
+  /**
+   * Reverse mode differentiation algorithm using a GPU
+   *
+   * Reference:
+   *
+   * Iain Murray: Differentiation of the Cholesky decomposition, 2016.
+   *
+   */
+  virtual void chain() {
+    using Eigen::MatrixXd;
+    MatrixXd Lbar_(M_, M_);
+    MatrixXd L_(M_, M_);
+    Lbar_.setZero();
+    L_.setZero();
+
+    size_t pos = 0;
+    for (size_type j = 0; j < M_; ++j) {
+      for (size_type i = j; i < M_; ++i) {
+        Lbar_.coeffRef(i, j) = variRefL_[pos]->adj_;
+        L_.coeffRef(i, j) = variRefL_[pos]->val_;
+        ++pos;
+      }
+    }
+
+    matrix_gpu L(L_);
+    matrix_gpu Lbar(Lbar_);
+    int block_size_
+        = M_ / opencl_context.tuning_opts().cholesky_rev_block_partition;
+    block_size_ = std::max(block_size_, 8);
+    block_size_ = std::min(
+        block_size_, opencl_context.tuning_opts().cholesky_rev_min_block_size);
+    // The following is a GPU implementation of
+    // the chain() function from the cholesky_block
+    // vari class implementation
+    for (int k = M_; k > 0; k -= block_size_) {
+      const int j = std::max(0, k - block_size_);
+      const int k_j_ind = k - j;
+      const int m_k_ind = M_ - k;
+
+      matrix_gpu R(k_j_ind, j);
+      matrix_gpu D(k_j_ind, k_j_ind);
+      matrix_gpu B(m_k_ind, j);
+      matrix_gpu C(m_k_ind, k_j_ind);
+
+      matrix_gpu Rbar(k_j_ind, j);
+      matrix_gpu Dbar(k_j_ind, k_j_ind);
+      matrix_gpu Bbar(m_k_ind, j);
+      matrix_gpu Cbar(m_k_ind, k_j_ind);
+
+      R.sub_block(L, j, 0, 0, 0, k_j_ind, j);
+      D.sub_block(L, j, j, 0, 0, k_j_ind, k_j_ind);
+      B.sub_block(L, k, 0, 0, 0, m_k_ind, j);
+      C.sub_block(L, k, j, 0, 0, m_k_ind, k_j_ind);
+
+      Rbar.sub_block(Lbar, j, 0, 0, 0, k_j_ind, j);
+      Dbar.sub_block(Lbar, j, j, 0, 0, k_j_ind, k_j_ind);
+      Bbar.sub_block(Lbar, k, 0, 0, 0, m_k_ind, j);
+      Cbar.sub_block(Lbar, k, j, 0, 0, m_k_ind, k_j_ind);
+
+      Cbar = Cbar * lower_triangular_inverse(D);
+      Bbar = Bbar - Cbar * R;
+      Dbar = Dbar - transpose(Cbar) * C;
+
+      // the implementation of the symbolic_rev inline function
+      // for the GPU
+      Dbar = transpose(D) * Dbar;
+      Dbar.triangular_transpose<TriangularMapGPU::LowerToUpper>();
+      D = transpose(lower_triangular_inverse(D));
+      Dbar = D * transpose(D * Dbar);
+      Dbar.triangular_transpose<TriangularMapGPU::LowerToUpper>();
+      // end of symbolic_rev inline function
+
+      Rbar = Rbar - transpose(Cbar) * B;
+      Rbar = Rbar - Dbar * R;
```

Comment:
seems like you could put this on the line above

Reply:
Done, thanks.